src/Nix/Effects/Basic.hs

```diff
@@ -64,13 +105,7 @@ defaultMakeAbsolutePath origPath = do
           Nothing -> getCurrentDirectory
           Just v  -> demand v $ \case
             NVPath s -> return $ takeDirectory s
-            v ->
-              throwError
-                $  ErrorCall
-                $  "when resolving relative path,"
-                ++ " __cur_file is in scope,"
-                ++ " but is not a path; it is: "
-                ++ show v
+            v -> throwError $ ErrorCall $ displayException $ CurFileIsntPath v
```

Comment:
You don't want to call `displayException` here, but where the `m` is eliminated.

Reply:
Do you mean to make a call out of the `do` block/blocks, or when `-> m FilePath` would be reduced? Surely I am missed something understanding your recommendation.

I would mark your point TODO right now, and dodge it to the further cycle over code. And while then maybe you would write or myself would get more understanding on the Exception process.

Since current approach can be drop-in mass-produced in the code I would first do this.

Reply:
Your suggestions helped me.

Reply:
Put most simply, there should be no `ErrorCall` in the code base. If you grep for `ErrorCall` you should get nothing.

Reply:
```haskell
-- | This is thrown when the user calls 'error'. The first @String@ is the
-- argument given to 'error', second @String@ is the location.
data ErrorCall = ErrorCallWithLocation String String
```
currently it seems to be the default constructor for such purposes (currently [`base GHC.Exception`](https://hackage.haskell.org/package/base-4.12.0.0/docs/src/GHC.Exception.html#line-66) considered the main path).

Reply:
Keep in mind throwError comes from the MonadError class and is used purely here (I think). There's no Haskell exceptions from this stuff.

Reply:
Ok. The Haskell vocabulary and situation is completely perplexing the communication here.
Probably you are saying that the process is pure, until the computation arrives outside into IO monad. Of course - strictly by theory that it is true, unpurity happens in IO monad.

`Throw` in combination with `error` terms themselves by the name state that it is ending not pure in any variant (Error/Exception or Async Exception/Sync Exception) of terminology I know of.
If throws are not caught - they go into IO monad in the end and trigger the effects. And the majority of situations of these exception constructors are not recoverable (unpure), and some are possible to recover with request of IO input (unpure), and the messages are there to be printed/logged, which you guessed...

Yes, the code inside IO monad is effectful and moves into C, for what I know. But if this pure functions are most directly exist to raise the effects - I think we can debate between views on pure/unpure terms indefinitely in this functions.

Reply:
But I just basically restated that this is that everyone knows.

From Haskell terms of pure realm - yes, `throwM` & `MonadThrow` are pure by themselves.
But again - these Exception situations are not recoverable, so they are resulting in IO.

Reply:
https://hackage.haskell.org/package/mtl-2.2.2/docs/Control-Monad-Except.html#v:throwError see the instances for that class. there is the IOException instance which uses exceptions but I don't think we are using that.

Reply:
The `throwError` is redefined in HNix:
https://github.com/haskell-nix/hnix/blob/c03a40aceac5221c82603c9208b381a76c72100f/src/Nix/Frames.hs#L53-L58

GHCI:
```haskell
*Nix> :l src/Nix.hs
...
*Nix> :i throwError
throwError ::
  (Framed e m, Exception s, Control.Monad.Catch.MonadThrow m) =>
  s -> m a
  	-- Defined at src/Nix/Frames.hs:55:1
```
`toException` [is](https://hackage.haskell.org/package/base-4.12.0.0/docs/src/GHC.Exception.Type.html#toException):
```haskell
class (Typeable e, Show e) => Exception e where
    toException   :: e -> SomeException
    fromException :: SomeException -> Maybe e

    toException = SomeException
    fromException (SomeException e) = cast e
```

& when `throwM` is launched - it shortens the monadic computation.

There is a bunch of custom HNix type classes & instances extending the IO monad in `src/Nix/Effects.hs`:
https://github.com/haskell-nix/hnix/blob/c03a40aceac5221c82603c9208b381a76c72100f/src/Nix/Effects.hs#L86-L113

`Effects.hs` has a second largest Exception counter over all HNix files, I am moving towards it. Also it is the root of all evil, so I decided to learn the simplier, more  parts before.

I just already learned a bunch of project code while working on it. Also normal tooling fitting to the project helps: https://github.com/Anton-Latukha/haskell-notes#integration-of-nixosnix-with-haskell-ide-engine-hie-and-emacs-spacemacs.

Reply:
Also I see (as also you can see here) there is too much forward and backward transformations of Exceptions, this computations can be substantially reduced.

Reply:
My apologies I didn't realize we were doing those things. @jwiegley has the final say, but I would ditch Exception and SomeException and only use closed error types. I really don't like open error types.

Reply:
No apologies, as I do the errors everyone does them.

Currently I'm just doing a bunch of Exception constructors as was asked.
I am gathering a list of major point questions to ask/tweak during the merge preparation.

Reply:
In that case you will want to implement the `Exception` class so you can `throwError $ YourNewNiceType` directly,

Reply:
@Ericson2314 is right, instead of `ErrorCall $ displayException $ EMonadExecMissingProgram u u`, it should just be `EMonadExecMissingProgram u u`. The rendering into a string part should only happen at reporting time.

Reply:
Yep, reduced that.